packages/grafana-ui/src/components/DateTimePickers/TimeRangeContext.tsx

```diff
@@ -0,0 +1,103 @@
+import { ReactNode, createContext, useCallback, useEffect, useMemo, useState, useContext } from 'react';
+
+import { TimeRange } from '@grafana/data';
+
+type TimeRangeContextValue = TimeRangeContextHookValue & {
+  // These are to be used internally and aren't passed to the users of the hook.
+
+  // Called when picker is mounted to update the picker count.
+  addPicker(): void;
+
+  // Called when picker is unmounted to update the picker count.
+  removePicker(): void;
+};
+
+export type TimeRangeContextHookValue = {
+  // If the time range is synced, this is the value that all pickers should show.
+  syncedValue?: TimeRange;
+
+  // If the time range is synced across all visible pickers.
+  synced: boolean;
+
+  // If it is possible to sync the time range. This is based just on the number of pickers that are visible so if
+  // there is only one, the sync button can be hidden.
+  syncPossible: boolean;
+
+  // Action passed to the picker to interact with the sync state.
+  // Sync the time range across all pickers with the provided value. Can be also used just to update a value when
+  // already synced.
+  sync(value: TimeRange): void;
+  unSync(): void;
+};
+
+const TimeRangeContext = createContext<TimeRangeContextValue | undefined>(undefined);
+
+export function TimeRangeProvider({ children }: { children: ReactNode }) {
+  // We simply keep the count of the pickers visible by letting them call the addPicker and removePicker functions.
+  const [pickersCount, setPickersCount] = useState(0);
+  const [syncedValue, setSyncedValue] = useState<TimeRange>();
+
+  const sync = useCallback((value: TimeRange) => {
+    setSyncedValue(value);
+  }, []);
+
+  const unSync = useCallback(() => {
+    setSyncedValue(undefined);
+  }, []);
```

Comment:
Can these two functions be now defined inline? Wonder why these two are wrapped in useCallback while add- and remove picker aren't 🤔 But in general it doesn't seem like there's much benefit to wrapping them in useCallback.

Reply:
good point. The initial idea was that because the functions don't have dependencies they don't have to ever change, so in case they are ever used as a dependency in other hook they won't affect their rerun but don't think that is a big deal in this case.